### ISSUE TITLE
Add inversion/eraser-end detection for tablet pens

### ISSUE THREAD

**Issue by hansemro** [godotengine/godot-proposals#4702]
### Describe the project you are working on

Lorien (https://github.com/mbrlabs/Lorien), which is an infinite-canvas drawing program that benefits from pen/wacom/stylus input method. This project uses Godot 3.4/3.x

### Describe the problem or limitation you are having in your project

I am currently working to resolve a feature request to map the eraser tool to the eraser-end of the pen (https://github.com/mbrlabs/Lorien/issues/118). Implementing this feature requires checking whether the pen is inverted or not. However, Godot engine does not currently provide enough information to determine which end of the pen is used and treats both the same.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add pen_inverted boolean property and getter/setter functions (under InputEventMouseMotion) similar to how pen pressure is polled.
- pen_inverted is true when eraser-end is used and false otherwise.
- `void set_pen_inverted(bool p_inverted)` :  requires platform-specific code to get status.
- `bool get_pen_inverted(void)` : used to poll inversion status

With this in place, my problem can be solved by polling inversion status and if-else statements to map a different tool on both ends of the pen.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

My proposal is based off my initial prototyping work: https://github.com/hansemro/godot/discussions/1

On a mouse-motion event, we leverage existing platform-specific APIs to determine and set the pen's inversion status. Once set, the event can be polled for pen_inverted property with `get_pen_inverted`. We do not expect game/project developers to use `set_pen_inverted`.
- Relevant API update commit: https://github.com/hansemro/godot/commit/a2b5742f806a5085fae81f93093b010f47d615e6

Most of the non-trivial work comes in handling platform-specific APIs. Below, I will cover some ideas that I have successfully tested in Godot `3.x` tree. For unhandled platforms, we can assume the pen to not be inverted.

# Windows

Both wintab and winink APIs have accessible properties to poll inversion status. Note that only wintab is expected to work under wine.

## wintab API

Check for TPS_INVERT flag in WTPacket.pkStatus.
- TPS_INVERT: https://github.com/Wacom-Developer/wacom-device-kit-windows/blob/d2bc78fe79d442a3d398f750357e46effbca1daa/Wintab%20CAD%20Test/SampleCode/WINTAB.H#L513
- pkStatus: https://github.com/Wacom-Developer/wacom-device-kit-windows/blob/d2bc78fe79d442a3d398f750357e46effbca1daa/Wintab%20CAD%20Test/SampleCode/PKTDEF.H#L170

Relevant prototype commit: https://github.com/hansemro/godot/commit/2af8a583795a8f6bfc8636331efb208e97574f6a

## winink API

Check for PEN_FLAG_INVERTED or PEN_FLAG_ERASER flags in POINTER_PEN_INFO.penFlags.
- pen flags: https://github.com/mingw-w64/mingw-w64/blob/35fa6bdbc14383c9078520492974ff5ab31ba796/mingw-w64-headers/include/winuser.h#L2679-L2680
- pen flags: https://docs.microsoft.com/en-us/windows/win32/inputmsg/pen-flags-constants
- POINTER_PEN_INFO: https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-pointer_pen_info

Relevant prototype commit: https://github.com/hansemro/godot/commit/75ca2cc74c57b7e4b53c31047cfb38a3acacece3

# Linux/X11

When input devices are polled, check if the word "eraser" is in the device's name. If it is present, mark the device in a device map as inverted. The strstr implementation is based off gtk/gdk's method for distinguishing PEN/ERASER input source: https://gitlab.com/gnome/gtk/-/blob/master/gdk/x11/gdkdevicemanager-xi.c#L235. Personally, I don't like this method, but I was not able to find a better way.

Relevant prototype commits:
- https://github.com/hansemro/godot/commit/a7f43e594145fb8fa5e03d74f93895b9a1cf0170
    - this is needed to get slave pointer id and name instead of `Virtual core pointer` with non-unique ids. Without this, all devices will have the same device id and name which will break this implementation.
    - when processing events, we can use sourceid to get slave pointer id to retrieve pen_inverted status for that device from a map.
- https://github.com/hansemro/godot/commit/2dff4067c2c7b577ee440d3f99a05479ecc3a85c

# macOS

TODO

# Other Platforms

TODO

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Although this feature may not be used often, it cannot easily be worked around as different platforms handle pen inversion differently.

### Is there a reason why this should be core and not an add-on in the asset library?

This feature will fill a hole in the already-existing pen/tablet/wacom support. This feature could also help distinguish Godot from other engines.

**Comment by Calinou** [comment]
cc @bruvzg

**Comment by hansemro** [comment]
Added macOS implementation although I cannot readily test this.

**Comment by hansemro** [comment]
Added simple demo project for testing: https://github.com/hansemro/InvertedPenGodotTest

Windows users can change between wintab and winink by setting the field in Project Settings->Display->Window->Tablet Driver

**Comment by mbrlabs** [comment]
To get more visibility on this i recommend opening a Proof-of-Concept PR (can be a draft) for this so people can test it with the build artifacts. You would also have to make a version for the master branch (Godot 4) eventually.

**Comment by hansemro** [comment]
> To get more visibility on this i recommend opening a Proof-of-Concept PR (can be a draft) for this so people can test it with the build artifacts. You would also have to make a version for the master branch (Godot 4) eventually.

Yeah, it looks like I need to get a PR for master branch first. Will work on it soon.

**Comment by akien-mga** [comment]
Implemented by https://github.com/godotengine/godot/pull/62212 for desktop platforms. I guess we should keep the proposal open to evaluate what's needed for Android/iOS too eventually.

**Comment by hansemro** [comment]
Android has inversion property, but I don't have a device to test with. https://source.android.com/devices/accessories/stylus

**Comment by hansemro** [comment]
Not currently possible in iOS, but it might be worth looking into double-tapping gesture on Apple Pencils.

**Comment by Calinou** [comment]
> Android has inversion property, but I don't have a device to test with. [source.android.com/devices/accessories/stylus](https://source.android.com/devices/accessories/stylus)

You should be able to connect a USB graphics tablet to an Android device using an USB OTG adapter, then enabling OTG mode in the device's advanced settings (at least with Wacom tablets).

**Comment by hansemro** [comment]
https://support.wacom.com/hc/en-us/articles/1500006342822-Which-pen-tablet-can-I-use-with-Android-

Don't have a supported tablet it seems. I might pick up CL-4000 to play around with on Android.

**Comment by hansemro** [comment]
I was able to test Intuos Pro M briefly on friend's Galaxy S21 and it actually treats eraser-end as an eraser in note-taking applications. I guess support really varies on android device used (I wonder if I can remedy this with a custom kernel). Depending on how easy it is to set up android development environment, I may revisit adding support for android soon.
